packages/go_router/lib/src/matching.dart

```diff
@@ -157,15 +178,24 @@ List<RouteMatch> _getLocRouteRecursively({
       continue;
     } else {
       // Otherwise, recurse
-      final String childRestLoc =
-          loc.substring(match.subloc.length + (match.subloc == '/' ? 0 : 1));
-      assert(loc.startsWith(match.subloc));
-      assert(restLoc.isNotEmpty);
+      final String childRestLoc;
+      final String newParentSubLoc;
+      if (match.route is ShellRoute) {
+        childRestLoc = restLoc;
+        newParentSubLoc = parentSubloc;
+      } else {
+        assert(loc.startsWith(match.location));
+        assert(restLoc.isNotEmpty);
+
+        childRestLoc = loc
+            .substring(match.location.length + (match.location == '/' ? 0 : 1));
+        newParentSubLoc = match.location;
+      }
 
       final List<RouteMatch> subRouteMatch = _getLocRouteRecursively(
         loc: loc,
         restLoc: childRestLoc,
-        parentSubloc: match.subloc,
+        parentSubloc: newParentSubLoc,
```

Comment:
we may as well remove all the mention of subLoc, it is also against flutter style guide to use abbreviation.

Reply:
Let's not rename things in this PR as per https://github.com/flutter/packages/pull/2453#discussion_r961143765